src/linear_algebra/projective_space/basic.lean

```diff
@@ -0,0 +1,232 @@
+    rintros u v ⟨a,ha⟩,
+    use a⁻¹,
+    rw [← ha, ← mul_smul, inv_mul_cancel, one_smul],
+    intros c, apply v.2,
+    simpa [c] using ha.symm,
+  end }
+
+
+/-- The projectivization of the `K`-vector space `V`. -/
+def projectivization := quotient (projectivization_setoid K V)
+
+namespace projectivization
+
+variables {V}
+
+/-- Construct an element of the projectivization from a nonzero vector. -/
+def mk (v : V) (hv : v ≠ 0) : projectivization K V := quotient.mk' ⟨v,hv⟩
+
+/-- A variant of `projectivization.mk` in terms of a subtype. `mk` is preferred. -/
+def mk' (v : { v : V // v ≠ 0 }) : projectivization K V := quotient.mk' v
+
+@[simp] lemma mk'_eq_mk (v : { v : V // v ≠ 0}) :
+  mk' K v = mk K v v.2 :=
+by { dsimp [mk, mk'], congr' 1, simp }
+
+noncomputable instance [nontrivial V] : inhabited (projectivization K V) :=
+let e := exists_ne (0 : V) in ⟨mk K e.some e.some_spec⟩
+
+variable {K}
+
+/-- Choose a representative of `v : projectivization K V` in `V`. -/
+protected noncomputable def rep (v : projectivization K V) : V := v.out'.1
+
+lemma rep_nonzero (v : projectivization K V) : v.rep ≠ 0 := v.out'.2
+
+@[simp]
+lemma mk_rep (v : projectivization K V) :
+  mk K v.rep v.rep_nonzero = v :=
+by { dsimp [mk, projectivization.rep], simp }
+
+open finite_dimensional
+
+/-- Consider an element of the projectivization as a submodule of `V`. -/
+protected def submodule (v : projectivization K V) : submodule K V := K ∙ v.rep
+
+variable (K)
+
+lemma exists_of_mk_eq_mk  (v w : V) (hv : v ≠ 0) (hw : w ≠ 0) (h : mk K v hv = mk K w hw) :
+  ∃ (a : K) (ha : a ≠ 0), a • v = w :=
+begin
+  obtain ⟨a,h⟩ := quotient.exact' h,
+  refine ⟨a,_,h⟩,
+  intros c, refine hw _,
+  simpa [c] using h.symm,
+end
+
+lemma exists_smul_eq_mk_rep
+  (v : V) (hv : v ≠ 0) : ∃ (a : K) (ha : a ≠ 0), a • v = (mk K v hv).rep :=
+exists_of_mk_eq_mk _ _ _ hv (rep_nonzero _) (by simp)
+
+variable {K}
+
+@[simp]
+lemma submodule_mk (v : V) (hv : v ≠ 0) : (mk K v hv).submodule = K ∙ v :=
+begin
+  dsimp only [projectivization.submodule],
+  obtain ⟨a, ha, h⟩ := exists_smul_eq_mk_rep K v hv,
+  rw ← h,
+  apply submodule.span_singleton_smul_eq _ ha,
+end
+
+lemma finrank_submodule (v : projectivization K V) : finrank K v.submodule = 1 :=
+finrank_span_singleton v.rep_nonzero
+
+instance (v : projectivization K V) : finite_dimensional K v.submodule :=
+by { dsimp [projectivization.submodule], apply_instance }
+
+lemma submodule_injective : function.injective
+  (projectivization.submodule : projectivization K V → submodule K V) :=
+begin
+  intros u v h, replace h := le_of_eq h,
+  erw submodule.le_span_singleton_iff at h,
+  rw [← mk_rep v, ← mk_rep u],
+  symmetry,
+  exact quotient.sound' (h u.rep (submodule.mem_span_singleton_self _))
+end
+
+variables (K V)
+/-- The equivalence between the projectivization and the
+collection of subspaces of dimension 1. -/
+noncomputable
+def equiv_submodule : projectivization K V ≃ { H : submodule K V // finrank K H = 1 } :=
```

Comment:
I feel like there's a construction for setoids to be extracted here.  The closest we have, not quite the same, is [setoid.quotient_ker_equiv_of_surjective](https://leanprover-community.github.io/mathlib_docs/data/setoid/basic.html#setoid.quotient_ker_equiv_of_surjective).

Reply:
What sort of result did you have in mind? If we had defined the relation for `projectivization_setoid` as the kernel of the map sending `v` to `K . v`, then we could use `quotient_ker_equiv...`. But with the current definition, I'm not sure if it's worthwhile to try to come up with some general construction for setoids that works in this case, if no one will ever use it!